python/ml4ir/base/pipeline.py

```diff
@@ -390,6 +390,26 @@ def run(self):
                 experiment_tracking_dict, orient="index", columns=["value"]
             ).to_csv()
 
+            # temperature scaling
+            if self.args.execution_mode in {
```

Comment:
@marjanhs Thinking out loud - should we do this right after training, before we evaluate? Does that make sense? Will the model saving work that way?

Reply:
It needs to be applied on a trained model. It does not matter its place after or before evaluation. Why do u thin it won't work if it moves to a place before evaluation ?

Reply:
For temperature scaling, I don't think it matters. It might for other techniques. Not sure at the moment. Let's leave it here for now. Looks good.

Reply:
@lastmansleeping @marjanhs we should discuss a bit more about this choice (perhaps in a follow up PR with an approved design before). My concerns: 
-  I think it is better to have a dedicated ExecutionModeKey for calibration. It can be misleading to have it as part of the model training for example. This because it is a process done after training completes, possibly requiring its very own data. Which leads me believe that a typical train/val/test split may not be enough and we may overfit in what we are reporting here.  
- What if we support 2 types of calibration and ones wants to compare for example? Would they need 2 model configs, each one repeating the other? Which suggests that the calibration requires a config, but preferably imho its own (clear concern separation) as it seems to be an independent method from the model itself. 
- Right now, it seems we are not saving the model with calibration just some predictions. This can also be misleading for the user and a bit concerning. These predictions are not copied in HDFS (or perhaps they are as part of the finish method of classification). 
- Worth discussing a bit how one would serve this in the scala part too.

Reply:
I also believe that due to these inconsistencies, we are better using it under an `experimental` name of something. But we may be far from this maturity

Reply:
Thank you @balikasg for the comments. 

- regarding supporting  different methods of calibration, they are a bit more complicated than temperature scaling. For the, we do need to discuss how to engage them in ml4ir. Currently it is part of [model_config](https://github.com/salesforce/ml4ir/pull/115/files#diff-1e0df01beb104b660c3c2805412d4527f3b35fe61e76ecce8d6410101e96fee9R27)

- I personally am not in favor  of adding a separate ExecutionModeKey as calibration can be a post processing step after training or on a trained model for evaluation step. 

- Saving the model with calibration was also my concern. I see two ways to do that : 1- only save temperature (scalar value) and then have a specific load function that joins the saved trained model with temperature 2- save the intermediate model by adding temperature . there might be other ways that worth to discuss (but won't add  for this PR). 

- about saving the scores in HDFS, should I added to the finish?

Reply:
(btw I will raise another PR to add it to the finish)

Reply:
I'm most interested in the model saving part. I am not comfortable with how we are doing it right now. But let's look at it in a follow up PR.

Reply:
I'm also ok with a followup item tracking the scaling as part of the model, but it does need to be done to make this fully functional.  So we should annotate this clearly in the docs (docstring and README), and then file a followup work item to track saving + serving this temperature scaling wrapper.

Reply:
@marjanhs we are saying the same thing I believe. All the options you lay down simply suggest that the user should explicitly say 'do-calibration', and the model-config is not ideal because it describes the model architecture. I see there is a consensus for reworking  the saving part, we can discuss there with any experience and feedback we gather from executing the code at scale.